src/compiler/utilities.ts

```diff
@@ -8201,60 +8203,2241 @@ function Symbol(this: Symbol, flags: SymbolFlags, name: __String) {
+        this.data.isolatedSignatureType = value;
+    }
+    get instantiations() {
+        return this.data.instantiations;
+    }
+    set instantiations(value: any) {
+        this.data.instantiations = value;
+    }
+    get implementationSignatureCache() {
+        return this.data.implementationSignatureCache;
+    }
+    set implementationSignatureCache(value: any) {
+        this.data.implementationSignatureCache = value;
+    }
+}
+
+// export const NodeDataImplInstantiations: Partial<Record<SyntaxKind, number>> = {}
+// class NodeDataImpl {
+//     // constructor(kind: SyntaxKind) {
+//     //     NodeDataImplInstantiations[kind] = (NodeDataImplInstantiations[kind] ?? 0) + 1;
+//     // }
+//     // original = undefined;
+//     // modifierFlagsCache = ModifierFlags.None
+// }
+// export const NodeImplInstantiations: Partial<Record<SyntaxKind, number>> = {}
+
+// export const allNodes = new Set<NodeImpl>();
+/** @internal */
+export class NodeImpl {
+    pos: number;
+    end: number;
+    kind: number;
+    id: number;
+    flags: NodeFlags;
+    transformFlags: TransformFlags;
+    parent: Node;
+    // emitNode: EmitNode | undefined;
+    modifierFlagsCache: ModifierFlags;
+    // original: undefined;
+    escapedText: undefined;
+    // jsDoc: any;
+    flowNode: any;
+    symbol: any;
+    __symbolTestOutputCache?: any; // For tests only
+    constructor(kind: SyntaxKind, pos: number, end: number) {
+        this.pos = pos;
+        this.end = end;
+        this.kind = kind;
+        this.id = 0;
+        this.flags = NodeFlags.None;
+        this.transformFlags = TransformFlags.None;
+        this.modifierFlagsCache = ModifierFlags.None;
+
+        this.parent = undefined!;
+        // this.emitNode = undefined;
+
+        // this.original = undefined;
+        this.escapedText = undefined;
+        // this.jsDoc = undefined;
+        this.flowNode = undefined;
+        this.symbol = undefined;
+        // allNodes.add(this);
+
+        if (Debug.isDebugging || tracing) {
+            this.__symbolTestOutputCache = undefined;
+        }
+    }
+    data: any = undefined;
+
+    get jsDoc() {
+        return this.data?.jsDoc;
+    }
+    set jsDoc(value: any) {
+        if (!this.data) {
+            this.data = { jsDoc: value };
+        }
+        else {
+            this.data.jsDoc = value;
+        }
+    }
+    get emitNode() {
+        return this.data?.emitNode;
+    }
+    set emitNode(value: EmitNode) {
+        if (!this.data) {
+            this.data = { emitNode: value };
+        }
+        else {
+            this.data.emitNode = value;
+        }
+    }
+
+    get original() {
+        return this.data?.original;
+    }
+    set original(value: any) {
+        if (!this.data) {
+            this.data = { original: value };
+        }
+        else {
+            this.data.original = value;
+        }
+    }
+
+    get sourceText() {
+        return this.data?.sourceText;
+    }

```

Comment:
It might look a bit silly, but if any of these accessors are a particular hot path, you can make an accessor monomorphic by having a switch/case on `this.kind` and having several branches that all read e.g. `this.data.sourceText` but behind different `case` headers.

Reply:
I tried that. It did not improve performance in my experiment. And I tried several different versions.. switch, if - linear search, if-binary search. All performed worse.